x/outline-fetch/main.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/transport/socks5"
+	"github.com/Jigsaw-Code/outline-sdk/transport/split"
+)
+
+func makeStreamDialer(transportConfig string) (transport.StreamDialer, error) {
+	if transportConfig == "" {
+		return &transport.TCPStreamDialer{}, nil
+	}
+
+	accessKeyURL, err := url.Parse(transportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse access key: %w", err)
+	}
+	switch accessKeyURL.Scheme {
```

Comment:
Shall we include this logic into the shared `ParseAccessKey`? So the function could return different configs:

```
type Scheme uint

const (
  Shadowsocks Scheme = 1 << iota
  Socks5
  Splitter
)

type Config interface {
}

type SSConfig interface {
  Config
}

type Socks5Config interface {
  Config
}

func ParseAccessKey(accessKey string, acceptedSchemes Scheme) (Scheme, Config) {
}
```

Also the shared `NewOutlineStreamDialer(accessKey string, acceptedSchemes Scheme)` may also parse to different types.

Reply:
The config parser should just output the dialer. No need for introducing intermediate configs.